From fcdf6a0c91511c0ff0f15026eba98a0624d8cb5f Mon Sep 17 00:00:00 2001
From: James Morse <james.morse@arm.com>
Date: Mon, 22 Jan 2024 14:58:06 +0800
Subject: [PATCH 654/960] downstream: arm_mpam: resctrl: Add boilerplate cpuhp
 and domain allocation

resctrl has its own data structures to describe its resources. We
can't use these directly as we play tricks with the 'MBA' resource,
picking the MPAM controls or monitors that best apply. We may export
the same component as both L3 and MBA.

Add mpam_resctrl_exports[] as the array of class->resctrl mappings we
are exporting, and add the cpuhp hooks that allocated and free the
resctrl domain structures.

While we're here, plumb in a few other obvious things.

Signed-off-by: James Morse <james.morse@arm.com>
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/platform/mpam/Makefile        |   2 +-
 drivers/platform/mpam/mpam_devices.c  |  12 ++
 drivers/platform/mpam/mpam_internal.h |  15 ++
 drivers/platform/mpam/mpam_resctrl.c  | 237 ++++++++++++++++++++++++++
 include/linux/arm_mpam.h              |   4 +
 5 files changed, 269 insertions(+), 1 deletion(-)
 create mode 100644 drivers/platform/mpam/mpam_resctrl.c

diff --git a/drivers/platform/mpam/Makefile b/drivers/platform/mpam/Makefile
index 8ad69bfa2aa2..37693be531c3 100644
--- a/drivers/platform/mpam/Makefile
+++ b/drivers/platform/mpam/Makefile
@@ -1 +1 @@
-obj-$(CONFIG_ARM_CPU_RESCTRL) += mpam_devices.o
+obj-$(CONFIG_ARM_CPU_RESCTRL) += mpam_devices.o mpam_resctrl.o
diff --git a/drivers/platform/mpam/mpam_devices.c b/drivers/platform/mpam/mpam_devices.c
index 8c1b3535be5e..cddec7ec228e 100644
--- a/drivers/platform/mpam/mpam_devices.c
+++ b/drivers/platform/mpam/mpam_devices.c
@@ -1410,6 +1410,9 @@ static int mpam_cpu_online(unsigned int cpu)
 	}
 	srcu_read_unlock(&mpam_srcu, idx);
 
+	if (mpam_is_enabled())
+		mpam_resctrl_online_cpu(cpu);
+
 	return 0;
 }
 
@@ -1469,6 +1472,9 @@ static int mpam_cpu_offline(unsigned int cpu)
 	}
 	srcu_read_unlock(&mpam_srcu, idx);
 
+	if (mpam_is_enabled())
+		mpam_resctrl_offline_cpu(cpu);
+
 	return 0;
 }
 
@@ -2140,6 +2146,12 @@ static void mpam_enable_once(void)
 	mutex_unlock(&mpam_list_lock);
 	cpus_read_unlock();
 
+	if (!err) {
+		err = mpam_resctrl_setup();
+		if (err)
+			pr_err("Failed to initialise resctrl: %d\n", err);
+	}
+
 	if (err) {
 		schedule_work(&mpam_broken_work);
 		return;
diff --git a/drivers/platform/mpam/mpam_internal.h b/drivers/platform/mpam/mpam_internal.h
index a7a80fa0a243..f82fc2d1fc0e 100644
--- a/drivers/platform/mpam/mpam_internal.h
+++ b/drivers/platform/mpam/mpam_internal.h
@@ -235,6 +235,16 @@ struct mpam_msc_ris {
 	struct msmon_mbwu_state	*mbwu_state;
 };
 
+struct mpam_resctrl_dom {
+	struct mpam_component	*comp;
+	struct rdt_domain	resctrl_dom;
+};
+
+struct mpam_resctrl_res {
+	struct mpam_class	*class;
+	struct rdt_resource	resctrl_res;
+};
+
 static inline int mpam_alloc_csu_mon(struct mpam_class *class)
 {
 	struct mpam_props *cprops = &class->props;
@@ -286,6 +296,11 @@ int mpam_msmon_read(struct mpam_component *comp, struct mon_cfg *ctx,
 		    enum mpam_device_features, u64 *val);
 void mpam_msmon_reset_mbwu(struct mpam_component *comp, struct mon_cfg *ctx);
 
+int mpam_resctrl_online_cpu(unsigned int cpu);
+int mpam_resctrl_offline_cpu(unsigned int cpu);
+
+int mpam_resctrl_setup(void);
+
 /*
  * MPAM MSCs have the following register layout. See:
  * Arm Architecture Reference Manual Supplement - Memory System Resource
diff --git a/drivers/platform/mpam/mpam_resctrl.c b/drivers/platform/mpam/mpam_resctrl.c
new file mode 100644
index 000000000000..b9c1292ff630
--- /dev/null
+++ b/drivers/platform/mpam/mpam_resctrl.c
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (C) 2021 Arm Ltd.
+
+#define pr_fmt(fmt) "mpam: resctrl: " fmt
+
+#include <linux/arm_mpam.h>
+#include <linux/cacheinfo.h>
+#include <linux/cpu.h>
+#include <linux/cpumask.h>
+#include <linux/errno.h>
+#include <linux/list.h>
+#include <linux/printk.h>
+#include <linux/rculist.h>
+#include <linux/resctrl.h>
+#include <linux/slab.h>
+#include <linux/types.h>
+
+#include <asm/mpam.h>
+
+#include "mpam_internal.h"
+
+/*
+ * The classes we've picked to map to resctrl resources.
+ * Class pointer may be NULL.
+ */
+static struct mpam_resctrl_res mpam_resctrl_exports[RDT_NUM_RESOURCES];
+
+static bool exposed_alloc_capable;
+static bool exposed_mon_capable;
+
+bool resctrl_arch_alloc_capable(void)
+{
+	return exposed_alloc_capable;
+}
+
+bool resctrl_arch_mon_capable(void)
+{
+	return exposed_mon_capable;
+}
+
+/*
+ * MSC may raise an error interrupt if it sees an out or range partid/pmg,
+ * and go on to truncate the value. Regardless of what the hardware supports,
+ * only the system wide safe value is safe to use.
+ */
+u32 resctrl_arch_get_num_closid(struct rdt_resource *ignored)
+{
+	return min((u32)mpam_partid_max + 1, (u32)RESCTRL_MAX_CLOSID);
+}
+
+struct rdt_resource *resctrl_arch_get_resource(enum resctrl_res_level l)
+{
+	if (l >= RDT_NUM_RESOURCES)
+		return NULL;
+
+	return &mpam_resctrl_exports[l].resctrl_res;
+}
+
+static int mpam_resctrl_resource_init(struct mpam_resctrl_res *res)
+{
+	/* TODO: initialise the resctrl resources */
+
+	return 0;
+}
+
+/* Called with the mpam classes lock held */
+int mpam_resctrl_setup(void)
+{
+	int err = 0;
+	struct mpam_resctrl_res *res;
+	enum resctrl_res_level i;
+
+	cpus_read_lock();
+	for (i = 0; i < RDT_NUM_RESOURCES; i++) {
+		res = &mpam_resctrl_exports[i];
+		INIT_LIST_HEAD(&res->resctrl_res.domains);
+		INIT_LIST_HEAD(&res->resctrl_res.evt_list);
+		res->resctrl_res.rid = i;
+	}
+
+	/* TODO: pick MPAM classes to map to resctrl resources */
+
+	for (i = 0; i < RDT_NUM_RESOURCES; i++) {
+		res = &mpam_resctrl_exports[i];
+		if (!res->class)
+			continue;	// dummy resource
+
+		err = mpam_resctrl_resource_init(res);
+		if (err)
+			break;
+	}
+	cpus_read_unlock();
+
+	if (!err && !exposed_alloc_capable && !exposed_mon_capable)
+		err = -EOPNOTSUPP;
+
+	if (!err) {
+		if (!is_power_of_2(mpam_pmg_max + 1)) {
+			/*
+			 * If not all the partid*pmg values are valid indexes,
+			 * resctrl may allocate pmg that don't exist. This
+			 * should cause an error interrupt.
+			 */
+			pr_warn("Number of PMG is not a power of 2! resctrl may misbehave");
+		}
+		err = resctrl_init();
+	}
+
+	return err;
+}
+
+static struct mpam_resctrl_dom *
+mpam_resctrl_alloc_domain(unsigned int cpu, struct mpam_resctrl_res *res)
+{
+	struct mpam_resctrl_dom *dom;
+	struct mpam_class *class = res->class;
+	struct mpam_component *comp_iter, *comp;
+
+	comp = NULL;
+	list_for_each_entry(comp_iter, &class->components, class_list) {
+		if (cpumask_test_cpu(cpu, &comp_iter->affinity)) {
+			comp = comp_iter;
+			break;
+		}
+	}
+
+	/* cpu with unknown exported component? */
+	if (WARN_ON_ONCE(!comp))
+		return ERR_PTR(-EINVAL);
+
+	dom = kzalloc_node(sizeof(*dom), GFP_KERNEL, cpu_to_node(cpu));
+	if (!dom)
+		return ERR_PTR(-ENOMEM);
+
+	dom->comp = comp;
+	INIT_LIST_HEAD(&dom->resctrl_dom.list);
+	dom->resctrl_dom.id = comp->comp_id;
+	cpumask_set_cpu(cpu, &dom->resctrl_dom.cpu_mask);
+
+	/* TODO: this list should be sorted */
+	list_add_tail(&dom->resctrl_dom.list, &res->resctrl_res.domains);
+
+	return dom;
+}
+
+/* Like resctrl_get_domain_from_cpu(), but for offline CPUs */
+static struct mpam_resctrl_dom *
+mpam_get_domain_from_cpu(int cpu, struct mpam_resctrl_res *res)
+{
+	struct rdt_domain *d;
+	struct mpam_resctrl_dom *dom;
+
+	lockdep_assert_cpus_held();
+
+	list_for_each_entry(d, &res->resctrl_res.domains, list) {
+		dom = container_of(d, struct mpam_resctrl_dom, resctrl_dom);
+
+		if (cpumask_test_cpu(cpu, &dom->comp->affinity))
+			return dom;
+	}
+
+	return NULL;
+}
+
+struct rdt_domain *resctrl_arch_find_domain(struct rdt_resource *r, int id)
+{
+	struct rdt_domain *d;
+	struct mpam_resctrl_dom *dom;
+
+	lockdep_assert_cpus_held();
+
+	list_for_each_entry(d, &r->domains, list) {
+		dom = container_of(d, struct mpam_resctrl_dom, resctrl_dom);
+		if (dom->comp->comp_id == id)
+			return &dom->resctrl_dom;
+	}
+
+	return NULL;
+}
+
+int mpam_resctrl_online_cpu(unsigned int cpu)
+{
+	int i;
+	struct mpam_resctrl_dom *dom;
+	struct mpam_resctrl_res *res;
+
+	for (i = 0; i < RDT_NUM_RESOURCES; i++) {
+		res = &mpam_resctrl_exports[i];
+
+		if (!res->class)
+			continue;	// dummy_resource;
+
+		dom = mpam_get_domain_from_cpu(cpu, res);
+		if (dom) {
+			cpumask_set_cpu(cpu, &dom->resctrl_dom.cpu_mask);
+			continue;
+		}
+
+		dom = mpam_resctrl_alloc_domain(cpu, res);
+		if (IS_ERR(dom))
+			return PTR_ERR(dom);
+	}
+
+	return 0;
+}
+
+int mpam_resctrl_offline_cpu(unsigned int cpu)
+{
+	int i;
+	struct rdt_domain *d;
+	struct mpam_resctrl_res *res;
+	struct mpam_resctrl_dom *dom;
+
+	for (i = 0; i < RDT_NUM_RESOURCES; i++) {
+		res = &mpam_resctrl_exports[i];
+
+		if (!res->class)
+			continue;	// dummy resource
+
+		d = resctrl_get_domain_from_cpu(cpu, &res->resctrl_res);
+		dom = container_of(d, struct mpam_resctrl_dom, resctrl_dom);
+
+		/* The last one standing was ahead of us... */
+		if (WARN_ON_ONCE(!d))
+			continue;
+
+		cpumask_clear_cpu(cpu, &d->cpu_mask);
+
+		if (!cpumask_empty(&d->cpu_mask))
+			continue;
+
+		list_del(&d->list);
+		kfree(dom);
+	}
+
+	return 0;
+}
diff --git a/include/linux/arm_mpam.h b/include/linux/arm_mpam.h
index 4ad8851c9e5e..a0f8d8e0102c 100644
--- a/include/linux/arm_mpam.h
+++ b/include/linux/arm_mpam.h
@@ -39,4 +39,8 @@ int mpam_register_requestor(u16 partid_max, u8 pmg_max);
 int mpam_ris_create(struct mpam_msc *msc, u8 ris_idx,
 		    enum mpam_class_types type, u8 class_id, int component_id);
 
+
+bool resctrl_arch_alloc_capable(void);
+bool resctrl_arch_mon_capable(void);
+
 #endif /* __LINUX_ARM_MPAM_H */
-- 
2.43.0

